packages/foundry/contracts/prediction-market/PredictionMarketHook.sol

```diff
@@ -0,0 +1,488 @@
+        uint256,
+        uint256[] memory,
+        uint256[] memory amountsOutRaw,
+        uint256[] memory,
+        bytes memory
+    ) public override returns (bool, uint256[] memory) {
+        _lastActivityBlockByPool[pool] = block.number;
+
+        return (true, amountsOutRaw);
+    }
+
+    /************************************
+     * Prediction Market Methods    
+     ************************************/ 
+
+    /**
+     * @notice Get the market id from the given market params
+     * @dev Useful for external callers
+     *
+     * @param pool Pool hosting the prediction market
+     * @param tokenA First token in the pair
+     * @param tokenB Second token in the pair
+     * @param closedAtTimestamp Timestamp for when the market closes
+     * @return marketId hashed market id
+     */
+    function getMarketId(
+        address pool,
+        address tokenA,
+        address tokenB,
+        uint256 closedAtTimestamp
+    ) public pure returns (bytes32 marketId) {
+        return PredictionMarketStorage.getMarketId(pool, tokenA, tokenB, closedAtTimestamp);
+    }
+
+    /**
+     * @notice Get a market given it's id
+     * @param id The market id
+     * @return market The resolved prediction market
+     */
+    function getMarketById(
+        bytes32 id
+    ) public view returns (PredictionMarket memory) {
+        return markets[id];
+    }
+
+    /**
+     * @notice Add liquidity to a given prediction market
+     * @dev The deposit token taken from the user balance will be the token0 of the given market, which is the 
+     * first token in the sorted pair
+     *
+     * User positions will be credited based on the current balance of "bets" between bull/bear outcomes. If the 
+     * market is uninitalized then the user will receive equal amounts of bull/bear units, representing a 50/50 
+     * probability of each outcome.
+     *
+     * The total value of a market at any given time is equal of the deposited liquidity. This is the amount that 
+     * will be split between the assets when the market is settled. For example, a market with 100 USDC deposited 
+     * and a 80/20 balance split between bull / bear would imply a bull price/probability of ($.8) and a bear 
+     * price/probability of ($.2). If this ratio continued to settlement then each bull unit would be worth $1.25 
+     * while each bear unit is worth $0
+     *
+     * One can think of this style of prediction market as an implementation of on-chain binary options 
+     *
+     * @param pool Pool hosting the prediction market
+     * @param tokenA First token in the pair
+     * @param tokenB Second token in the pair
+     * @param closedAtTimestamp Timestamp for when the market closes
+     * @param amount Deposit amount
+     * @param side Side to add liquidity to, optionally choose both to add proportionally
+     * @return position Resulting user liquidity position after deposit
+     */
+    function addLiquidity(
```

Comment:
👍 separating initializeMarket from addLiquidity

Having multiple markets open at once or not is something I went back and forth on, and I could see the validity of alternate implementations like you are suggesting. Users creating infinite markets is not ideal because it reduces the liquidity depth in each market and likely makes a UI on top of this harder to manage, while on the other side having 1 market open at a time likely wouldn't meet needs of users looking to trade on different time frames.

If i were to rewrite this, I would likely implement a minimum spacing between markets (i.e. 5 minutes) to prevent the proliferation of markets while supporting various time horizons.